src/sys/windows/named_pipe.rs

```diff
@@ -0,0 +1,610 @@
+    /// set" of initial configuration options.
+    pub fn new<A: AsRef<OsStr>>(
+        addr: A,
+        registry: &Registry,
+    ) -> io::Result<NamedPipe> {
+        let pipe = pipe::NamedPipe::new(addr)?;
+        NamedPipe::from_raw_handle(pipe.into_raw_handle(), registry)
+    }
+
+    /// TODO: Dox
+    pub fn from_raw_handle(
+        handle: RawHandle,
+        registry: &Registry,
+    ) -> io::Result<NamedPipe> {
+        // Generate a token
+        let token = NEXT_TOKEN.fetch_add(2, Relaxed) + 2;
+
+        // Create the pipe
+        let pipe = NamedPipe {
+            inner: Arc::new(Inner {
+                // Safety: not really unsafe
+                handle: unsafe { pipe::NamedPipe::from_raw_handle(handle) },
+                // transmutes to straddle winapi versions (mio 0.6 is on an
+                // older winapi)
+                connect: Overlapped::new(connect_done),
+                connecting: AtomicBool::new(false),
+                read: Overlapped::new(read_done),
+                write: Overlapped::new(write_done),
+                io: Mutex::new(Io {
+                    read: State::None,
+                    read_waker: None,
+                    write: State::None,
+                    write_waker: None,
+                    connect_error: None,
+                }),
+                pool: Mutex::new(BufferPool::with_capacity(2)),
+            }),
+        };
+
+        // Register the handle w/ the IOCP handle
+        poll::selector(registry)
+            .inner
+            .cp
+            .add_handle(usize::from(token), &pipe.inner.handle)?;
+
+        // Queue the initial read
+        Inner::post_register(&pipe.inner);
+
+        Ok(pipe)
+    }
+
+    /// Attempts to call `ConnectNamedPipe`, if possible.
+    ///
+    /// This function will attempt to connect this pipe to a client in an
+    /// asynchronous fashion. If the function immediately establishes a
+    /// connection to a client then `Ok(())` is returned. Otherwise if a
+    /// connection attempt was issued and is now in progress then a "would
+    /// block" error is returned.
+    ///
+    /// When the connection is finished then this object will be flagged as
+    /// being ready for a write, or otherwise in the writable state.
+    ///
+    /// # Errors
+    ///
+    /// This function will return a "would block" error if the pipe has not yet
+    /// been registered with an event loop, if the connection operation has
+    /// previously been issued but has not yet completed, or if the connect
+    /// itself was issued and didn't finish immediately.
+    ///
+    /// Normal I/O errors from the call to `ConnectNamedPipe` are returned
+    /// immediately.
+    pub fn connect(&self) -> io::Result<()> {
+        // "Acquire the connecting lock" or otherwise just make sure we're the
+        // only operation that's using the `connect` overlapped instance.
+        if self.inner.connecting.swap(true, SeqCst) {
```

Comment:
Nit: this can be lowered to `AcqRel`.

Reply:
This is not obviously true to me. How would this interact w/ the stores?